README.md
<!-- 
This README describes the package. If you publish this package to pub.dev,
this README's contents appear on the landing page for your package.

For information about how to write a good package README, see the guide for
[writing package pages](https://dart.dev/guides/libraries/writing-package-pages). 

For general information about developing packages, see the Dart guide for
[creating packages](https://dart.dev/guides/libraries/create-library-packages)
and the Flutter guide for
[developing packages and plugins](https://flutter.dev/developing-packages). 
-->

# CreaRadioButton - an easy radio button group
A package to create simple radio buttons.

## Example
### Change border on slected
![alt text](https://github.com/dutkaD/crea_radio/blob/main/assets/output.gif "Example options no border")
### Change background color on selected
![alt text](https://github.com/dutkaD/crea_radio/blob/main/assets/output2.gif "example options with border")
### Unselect is enabled
![alt text](https://github.com/dutkaD/crea_radio/blob/main/assets/unselect.gif "example options unselect")

## Features

Render a radio button group:
* with any number of buttons
* vertical or horizontal
* square or round

## Limitations
Currently only supports String as value and button label. An option for RadioGroup is passed via class RadioOption.

## Getting started
To use this plugin, add crea_radio_button as a dependency in your pubspec.yaml file.

## RadioButtonGroup

```dart 
String radioButtonInput = "";

 RadioButtonGroup(
                buttonHeight: 35,
                buttonWidth: 60,
                circular: true,
                mainColor: Colors.grey,
                selectedColor: Colors.pink.shade400,
                preSelectedIdx: defaultOptionIndex,
                options: [
                  RadioOption("10.0", "10"),
                  RadioOption("25.0", "25"),
                  RadioOption("50.0", "50"),
                  RadioOption("100.0", "100")
                ],
                callback: (RadioOption val) {
                  setState(() {
                    selectedOption = val;
                  });
                })


```

### Default selection
By default none of the options is preselected. You can specify index of the option you want to be preselected/
`defaultIdx: 0`

### Add Border for seleted option
By default there is no Border on a button. `selectedBorderSide can be used to add border for the selected option.

### Callback
Callback function returns the selected `RadioOption`


```dart 
String radioButtonInput = "";

RadioButtonGroup(
                      vertical: true,
                      options: [
                        RadioOption("1", "1"),
                        RadioOption("2", "2"),
                        RadioOption("3", "3")
                      ],
                      buttonHeight: 50,
                      buttonWidth: 50,
                      defaultIdx: defaultRadioIdx,
                      callback: (val) {
                        setState(() {
                          radioButtonInput = val.text;
                        });
                      });


```

## Additional information
Contribute here: https://github.com/dutkaD/crea_radio


CHANGELOG.md
## 0.5.0
* Unselect is enabled

## 0.4.0
* Change radio option Object to Generic

## 0.3.0
* Multirows and multicolumns

## 0.2.0
* Alignment (main axis, cross axis)
* Text style of the button can be changed
* BorderSide for selected option

## 0.1.1 - 0.1.4
* Fixes and refactorings

## 0.1.0
* `RadioOption` accepts any object as it's value
* `RadioOption` parameter `text` is renamed to `label`
* Updated example


## 0.0.2
* Improved documentation and added example

## 0.0.1
* A very simple radio button group with adjustable shape and colors.

test/crea_radio_button_test.dart
import 'package:crea_radio_button/crea_radio_button.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:flutter/material.dart';

void main() {
  RadioButtonGroup simpleRadio = RadioButtonGroup(
      options: [RadioOption("yes", "yes"), RadioOption("no", "no")],
      callback: (val) {
        print(val);
      });

  RadioButtonGroup simpleVerticalRadio = simpleRadio.copyWith(vertical: true);

  testWidgets('Vertical Radio Group builds a Column with two options',
      (tester) async {
    await tester.pumpWidget(Directionality(
      textDirection: TextDirection.ltr,
      child: simpleRadio,
    ));

    expect(find.byType(Row), findsOneWidget);
    expect(find.byType(Column), findsNothing);
    expect(find.byType(ElevatedButton), findsNWidgets(2));
  });

  testWidgets('Horizontal Radio Group builds a Row with two options',
      (tester) async {
    await tester.pumpWidget(Directionality(
        textDirection: TextDirection.ltr, child: simpleVerticalRadio));

    expect(find.byType(Row), findsNothing);
    expect(find.byType(Column), findsOneWidget);
    expect(find.byType(ElevatedButton), findsNWidgets(2));
  });
}


lib/crea_radio_button.dart
library crea_radio_button;

import 'package:flutter/material.dart';

/// [RadioOption] - an option passed to the [RadioButtonGroup]
class RadioOption<T> {
  /// an actual value of the option
  final T value;

  /// label of the option -> is rendered on the radio button
  final String label;

  RadioOption(this.value, this.label);
}

// class RadioValue<T> {
//   const RadioValue({
//     required this.value,
//   });
//   /// The value represented by this radio button.
//   final T value;
// }

/// Group of radio buttons, where only one option can be selected
/// Requires list of [options] and [callback] function
class RadioButtonGroup extends StatefulWidget {
  /// List of options for the group
  final List<RadioOption> options;

  /// Callback function
  final Function callback;

  /// How many rows or columns should be options devided into
  /// Default: 1. Options are places in a single column or row
  /// Should be > 0
  final int multilineNumber;

  /// Color which is used when the button is active (selected)
  final Color? selectedColor;

  /// Button text style
  final TextStyle? textStyle;

  /// Color which is used when the button is NOT selected
  final Color mainColor;

  /// Main Axis Alignment of the buttons. Default [MainAxisAlignment.spaceAround]
  final MainAxisAlignment? mainAxisAlignment;

  /// Cross Axis Alignment of the buttons. Default [CrossAxisAlignment.center]
  final CrossAxisAlignment? crossAxisAlignment;

  /// Direction of the [RadioButtonGroup].
  /// Default: vertical = false -> horizontal display
  /// Set vertical to true to display options as a column
  final bool vertical;

  /// Height of the option button in the group
  final double buttonHeight;

  /// Width of the option button in the group
  final double buttonWidth;

  /// Width between columns or rows of radio options
  final double betweenMultiLines;

  /// Space between buttons
  /// Default value:  8.0
  /// If [vertical] = true -> adds space vertically
  /// If [vertical] = false -> adds space horizontally
  final double spaceBetween;

  /// Index of the [RadioOption] in [options] that is selected by default
  final int? preSelectedIdx;

  /// Shape of the button
  /// Default [circular] = false -> Rectangular shape
  /// Set to true to have round buttons
  final bool circular;

  /// When pressing the selected button, the value of the group will be set to null
  /// Default: false
  final bool unselectEnabled;

  /// Defines the border for the selected button.
  /// Transparent and 0 with by default.
  final BorderSide? selectedBorderSide;

  const RadioButtonGroup(
      {Key? key,
      required this.options,
      required this.callback,
      this.selectedColor,
      this.mainColor = Colors.grey,
      this.vertical = false,
      this.buttonHeight = 50,
      this.buttonWidth = 50,
      this.circular = false,
      this.spaceBetween = 8.0,
      this.preSelectedIdx,
      this.mainAxisAlignment,
      this.crossAxisAlignment,
      this.textStyle,
      this.selectedBorderSide,
      this.multilineNumber = 1,
      this.betweenMultiLines = 0,
      this.unselectEnabled = false})
      : super(key: key);

  @override
  State<RadioButtonGroup> createState() => _RadioButtonGroupState();

  RadioButtonGroup copyWith(
      {List<RadioOption>? options,
      Function? callback,
      Color? selectedColor,
      Color? mainColor,
      bool? vertical,
      double? buttonHeight,
      double? buttonWidth,
      double? spaceBetween,
      double? betweenMultiLines,
      MainAxisAlignment? mainAxisAlignment,
      CrossAxisAlignment? crossAxisAlignment,
      TextStyle? textStyle,
      int? preSelectedIdx,
      bool? circular,
      BorderSide? borderSide,
      bool? allowUnSelect}) {
    return RadioButtonGroup(
      options: options ?? this.options,
      callback: callback ?? this.callback,
      selectedColor: selectedColor ?? this.selectedColor,
      mainColor: mainColor ?? this.mainColor,
      vertical: vertical ?? this.vertical,
      buttonHeight: buttonHeight ?? this.buttonHeight,
      buttonWidth: buttonWidth ?? this.buttonWidth,
      spaceBetween: spaceBetween ?? this.spaceBetween,
      preSelectedIdx: preSelectedIdx ?? this.preSelectedIdx,
      circular: circular ?? this.circular,
      mainAxisAlignment: mainAxisAlignment ?? this.mainAxisAlignment,
      crossAxisAlignment: crossAxisAlignment ?? this.crossAxisAlignment,
      textStyle: textStyle ?? this.textStyle,
      selectedBorderSide: borderSide ?? this.selectedBorderSide,
      betweenMultiLines: betweenMultiLines ?? this.betweenMultiLines,
      unselectEnabled: allowUnSelect ?? this.unselectEnabled,
    );
  }
}

class _RadioButtonGroupState extends State<RadioButtonGroup> {
  Object? selectedInGroup;

  @override
  void initState() {
    var selectedIdx = widget.preSelectedIdx;
    if (selectedIdx != null) {
      selectedInGroup = widget.options[selectedIdx].value;
    }
  }

  List<Widget> getRadioButtons() {
    List<Widget> radioButtons = [];

    for (RadioOption radioOption in widget.options) {
      radioButtons.add(Padding(
        padding: EdgeInsets.symmetric(
            vertical: widget.vertical ? widget.spaceBetween : 0.0,
            horizontal: widget.vertical ? 0.0 : widget.spaceBetween),
        child: _RadioButton(
          buttonHeight: widget.buttonHeight,
          buttonWidth: widget.buttonWidth,
          colorSelected: widget.selectedColor,
          colorMain: widget.mainColor,
          value: radioOption.value,
          groupValue: selectedInGroup,
          text: radioOption.label,
          unselectEnabled: widget.unselectEnabled,
          onChanged: (value) {
            setState(() {
              selectedInGroup = value;
              if (value == null) {
                widget.callback(RadioOption(null, ""));
              } else {
                widget.callback(radioOption);
              }
            });
          },
          circular: widget.circular,
          textStyle: widget.textStyle,
          borderSide: widget.selectedBorderSide,
        ),
      ));
    }
    return radioButtons;
  }

  @override
  Widget build(BuildContext context) {
    List<Widget> buttons = getRadioButtons();
    return widget.vertical
        ? Row(
            children: [
              Column(
                mainAxisAlignment:
                    widget.mainAxisAlignment ?? MainAxisAlignment.spaceAround,
                crossAxisAlignment:
                    widget.crossAxisAlignment ?? CrossAxisAlignment.center,
                children: buttons,
              ),
            ],
          )
        : Column(
            children: getRows(widget.multilineNumber, buttons),
          );
  }

  List<Widget> getRows(int number, List<Widget> buttons) {
    if (number == 1) {
      return [
        Row(
            mainAxisAlignment:
                widget.mainAxisAlignment ?? MainAxisAlignment.spaceAround,
            crossAxisAlignment:
                widget.crossAxisAlignment ?? CrossAxisAlignment.center,
            children: buttons)
      ];
    } else {
      List<Widget> rows = [];
      int numPerRow = (buttons.length / number).ceil();
      int start = 0;
      for (var i = 0; i < number; i += 1) {
        rows.add(Padding(
          padding: EdgeInsets.symmetric(vertical: widget.betweenMultiLines / 2),
          child: Row(
            mainAxisAlignment:
                widget.mainAxisAlignment ?? MainAxisAlignment.spaceAround,
            crossAxisAlignment:
                widget.crossAxisAlignment ?? CrossAxisAlignment.center,
            children: buttons.sublist(
                start,
                start + numPerRow > buttons.length
                    ? buttons.length
                    : start + numPerRow),
          ),
        ));
        start = start + numPerRow;
      }
      return rows;
    }
  }
}

class _RadioButton extends StatelessWidget {
  final Object value;
  final Object? groupValue;
  final String text;
  final Function onChanged;
  final Color? colorSelected;
  final Color colorMain;
  final double buttonHeight;
  final double buttonWidth;
  final bool circular;
  final TextStyle? textStyle;
  final BorderSide? borderSide;
  final bool unselectEnabled;

  const _RadioButton({
    Key? key,
    required this.value,
    required this.groupValue,
    required this.text,
    required this.onChanged,
    required this.colorSelected,
    required this.colorMain,
    required this.buttonHeight,
    required this.buttonWidth,
    required this.circular,
    this.textStyle,
    this.borderSide,
    required this.unselectEnabled,
  }) : super(key: key);

  @override
  Widget build(BuildContext context) {
    final isSelected = value == groupValue;

    return _BasicButton(
      radius: circular ? 100 : 2.5,
      height: buttonHeight,
      width: buttonWidth,
      callback: () {
        if (isSelected && unselectEnabled) {
          onChanged(null);
        } else {
          onChanged(value);
        }
      },
      label: text,
      textStyle: textStyle,
      isSelected: isSelected,
      color: isSelected ? colorSelected : colorMain,
      borderSide: borderSide,
    );
  }
}

class _BasicButton extends StatelessWidget {
  final Function callback;
  final String label;
  final double height;
  final double width;
  final bool isSelected;
  final double radius;
  final Color? color;
  final BorderSide? borderSide;
  final TextStyle? textStyle;

  const _BasicButton(
      {Key? key,
      required this.callback,
      required this.label,
      required this.height,
      required this.width,
      this.radius = 0,
      this.color,
      this.textStyle,
      this.borderSide,
      required this.isSelected})
      : super(key: key);

  @override
  Widget build(BuildContext context) {
    return SizedBox(
      height: height,
      width: width,
      child: OutlinedButton(
        style: OutlinedButton.styleFrom(
          shape:
              RoundedRectangleBorder(borderRadius: BorderRadius.circular(20.0)),
          backgroundColor: color ?? Theme.of(context).primaryColor,
          side: borderSide == null || !isSelected
              ? const BorderSide(
                  color: Colors.transparent,
                )
              : borderSide,
        ),
        onPressed: () {
          callback();
        },
        child: Center(
          child: Text(
            label,
            style:
                textStyle ?? const TextStyle(color: Colors.white, fontSize: 16),
          ),
        ),
      ),
    );
  }
}


lib/main.dart
import 'package:crea_radio_button/crea_radio_button.dart';
import 'package:flutter/material.dart';

class AmountOption {
  String currencySign;
  String currencyName;
  double amount;

  AmountOption(this.currencySign, this.currencyName, this.amount);
}

void main() {
  runApp(MaterialApp(
    home: Scaffold(
      body: Center(
        child: Column(
          children: const [FrequencyContainer(), AmountContainer()],
        ),
      ),
    ),
  ));
}

class FrequencyContainer extends StatefulWidget {
  const FrequencyContainer({
    Key? key,
  }) : super(key: key);

  @override
  State<FrequencyContainer> createState() => _FrequencyContainerState();
}

class _FrequencyContainerState extends State<FrequencyContainer> {
  String label = "";

  List<RadioOption> options = [
    RadioOption("MONTHLY", "Monthly"),
    RadioOption("YEARLY", "Yearly")
  ];

  @override
  void initState() {
    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    return Container(
      height: 250,
      width: 400,
      decoration: BoxDecoration(
          borderRadius: BorderRadius.circular(20),
          border: Border.all(color: Colors.grey.shade200),
          color: Colors.grey.shade50),
      child: Padding(
        padding: const EdgeInsets.all(30.0),
        child: Column(
          children: [
            const Text(
              "How do you want to pay?",
              style: TextStyle(fontSize: 30),
            ),
            Padding(
              padding: const EdgeInsets.all(8.0),
              child: Text(
                label,
                style: const TextStyle(
                    fontSize: 25,
                    fontWeight: FontWeight.bold,
                    color: Colors.pink),
              ),
            ),
            const SizedBox(
              height: 25,
            ),
            RadioButtonGroup(
                options: options,
                unselectEnabled: true,
                textStyle: const TextStyle(fontSize: 15, color: Colors.white),
                mainAxisAlignment: MainAxisAlignment.center,
                selectedColor: Colors.grey.shade400,
                mainColor: Colors.grey.shade400,
                selectedBorderSide:
                    BorderSide(width: 2, color: Colors.pink.shade300),
                buttonWidth: 120,
                buttonHeight: 35,
                callback: (RadioOption val) {
                  setState(() {
                    label = val.label;
                  });
                  print(val);
                }),
          ],
        ),
      ),
    );
  }
}

class AmountContainer extends StatefulWidget {
  const AmountContainer({
    Key? key,
  }) : super(key: key);

  @override
  State<AmountContainer> createState() => _AmountContainerState();
}

class _AmountContainerState extends State<AmountContainer> {
  final int defaultOptionIndex = 0;
  late AmountOption? selectedOption;
  List<RadioOption> options = [
    RadioOption(AmountOption("\$", "EURO", 10.00), "10"),
    RadioOption(AmountOption("\$", "EURO", 25.00), "25"),
    RadioOption(AmountOption("\$", "EURO", 50.00), "50"),
    RadioOption(AmountOption("\$", "EURO", 100.00), "100"),
  ];

  @override
  void initState() {
    super.initState();
    selectedOption = options[defaultOptionIndex].value;
  }

  @override
  Widget build(BuildContext context) {
    return Container(
      height: 200,
      width: 400,
      decoration: BoxDecoration(
          borderRadius: BorderRadius.circular(20),
          border: Border.all(color: Colors.grey.shade200),
          color: Colors.grey.shade200),
      child: Padding(
        padding: const EdgeInsets.all(30.0),
        child: Column(
          children: [
            const Text("Select your amount"),
            Padding(
              padding: const EdgeInsets.all(8.0),
              child: Text(
                '\$ ${selectedOption?.amount ?? ""}',
                style:
                    TextStyle(fontSize: 20, color: Colors.pinkAccent.shade200),
              ),
            ),
            const SizedBox(
              height: 25,
            ),
            RadioButtonGroup(
                buttonHeight: 35,
                buttonWidth: 60,
                circular: true,
                mainColor: Colors.grey,
                selectedColor: Colors.pink.shade400,
                preSelectedIdx: defaultOptionIndex,
                options: options,
                callback: (RadioOption val) {
                  setState(() {
                    selectedOption = val.value;
                  });
                })
          ],
        ),
      ),
    );
  }
}
